### PR TITLE
DOC: rework documents and silence warnings during sphinx build

### PR DESCRIPTION
**Pull request by mattip**
This started with the first two commits which make the devdocs more prominent and add the howto_build and howto_docuement to the public document tree.

I then noticed all the warnings produced during documentation generation. It seems that the plots in `random` require scipy.special to produce the theoretical distribution alongside the sampled distribution.
There are still many more warnings, see the [circleci build](https://circleci.com/gh/numpy/numpy/958?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

Should I split this into separate smaller pull requests? I still would like to silence more of the warnings, so this is still a WIP

### CODE REVIEW

**Comment by rgommers** [comment]
These changes are very welcome, thanks!

Installing `scipy` for the doc build is fine with me. We also require `matplotlib`; dependencies for building docs aren't a big deal imho.

Splitting in multiple PRs isn't strictly necessary, but this can be merged now. Maybe make a difference between changes like adding a dependency, and keeping large diffs due to warning fixes in a separate PR?

File: doc/source/docs/howto_build_docs.rst [hunk 1]
```diff
@@ -38,7 +39,9 @@ that the correct version is imported by
 Note that you can eg. install NumPy to a temporary location and set
 the PYTHONPATH environment variable appropriately.
 
-After NumPy is installed, write::
+After NumPy is installed, install scipy since some of the plots in the random
+module require `scipy.scpecial` to display properly. Now you are ready to
```
Comment (rgommers) [inline doc/source/docs/howto_build_docs.rst]: typo `scpecial`

**Review by rgommers — CHANGES_REQUESTED** [review]
The `HOWTO_DOCUMENT` is linked to directly in many places, so needs a redirect. The `HOWTO_BUILD` doesn't I think. Other than that, this looks good.

**Comment by njsmith** [comment]
It would be great if we can eventually turn on `-W`, sphinx's convert-warnings-into-errors mode, since sphinx's idea of a "warning" is universally actually a bug. Doesn't have to be this PR though, I just wanted to mention its existence since I only recently learned about it :-)

**Comment by mattip** [comment]
> turn on -W, sphinx's convert-warnings-into-errors mode

Is it on for docstring tests? Some of the warnings are from there and not specific to documentation

File: doc/source/docs/howto_build_docs.rst (hunk 1 above)
Comment (mattip) [reply doc/source/docs/howto_build_docs.rst]: fixed

**Comment by mattip** [comment]
> The HOWTO_DOCUMENT is linked to directly in many places, so needs a redirect

Some of those are actually in the `numpy/numpydoc` repo. See [pr to fix it](https://github.com/numpy/numpydoc/pull/167)

**Comment by pv** [comment]
Maybe also edit doc/HOWTO_DOCUMENT.rst.txt so that it doesn't duplicate
the text in numpydoc? (Otherwise, they'll get out of sync...)

**Comment by mattip** [comment]
wat? This seems to be failing on python3.6.3 linux and windows, but I cannot reproduce
```
import tokenize, ast

with tokenize.open('numpy/distutils/misc_util.py') as fid:
    txt = ast.parse(fid.read())

```
I changed  `*` to `\*` in the docstrings so they are rendered as `*` by the restructured text. It should be ignored by ast.parse

**Comment by mattip** [comment]
@pv good point. They are already out of sync, comparing `doc/HOWTO_DOCUMENT.rst.txt` and (after submodule initialization) `doc/sphinxext/doc/format.rst'. I will remove the former in favor of the latter

**Comment by stefanv** [comment]
I read Nathaniel's comment to mean that the HOWTO_DOCUMENT is linked to externally by a bunch of websites and projects, so we cannot remove that document without putting some kind of redirect in place.  Just want to make sure that issue has been addressed.

**Comment by mattip** [comment]
ok, I added a stub document for random places like [this one](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/index.html) (search for the `numpy` link on that page) that link there

File: doc/source/docs/howto_build_docs.rst [hunk 2]
```diff
@@ -38,7 +39,9 @@ that the correct version is imported by
 Note that you can eg. install NumPy to a temporary location and set
 the PYTHONPATH environment variable appropriately.
 
-After NumPy is installed, write::
+After NumPy is installed, install scipy since some of the plots in the random
```
Comment (stefanv) [inline doc/source/docs/howto_build_docs.rst]: Inconsistent NumPy / SciPy spelling

File: doc/source/docs/index.rst [hunk 3]
```diff
@@ -0,0 +1,11 @@
+.. _documentation:
+
+Numpy's Documentation
```
Comment (stefanv) [inline doc/source/docs/index.rst]: NumPy capitalization

File: numpy/doc/glossary.py [hunk 4]
```diff
@@ -62,6 +62,12 @@
          >>> x.shape
          (3,)
 
+   big-endian
+       When storing a multi-byte value in memory as a sequence of bytes. the
```
Comment (stefanv) [inline numpy/doc/glossary.py]: Capitalization of t of "the"

**Review by stefanv — COMMENTED** [review]
This looks good to me.  I highlighted some minor typos, which may have existed before this PR.

File: numpy/doc/glossary.py (hunk 4 above)
Comment (mattip) [reply numpy/doc/glossary.py]: s/./,/   fixed

File: doc/source/docs/index.rst (hunk 3 above)
Comment (mattip) [reply doc/source/docs/index.rst]: fixed

File: doc/source/docs/howto_build_docs.rst (hunk 2 above)
Comment (mattip) [reply doc/source/docs/howto_build_docs.rst]: fixed

**Comment by mattip** [comment]
Turns out the numpydoc submodule was pinned to `e161379f02efc4` which was 0.7, from June2017

**Comment by charris** [comment]
Maybe. I'll try ye olde close and reopen trick.

**Comment by charris** [comment]
I don't know it the failing tests matter. But I suppose you could close this, and make another PR from the same branch just to see what happens,

**Comment by mattip** [comment]
The LGTM failure to import submodule numpydoc 0.8.0 which was released March 31 is a false positive. But merging to master would mean tests fail on master, reverting to 0.7.0 which was the version before I started touching means we loose the transition of `HOWTO_DOCUMENT.rst.txt` in numpy/master to the documentation in numpydoc, which was added as part of 0.7 -> 0.8

Maybe if we wait a week it will resolve itself?

**Comment by pv** [comment]
Does the lgtm.com failure matter for master? iiuc, it only looks
whether if the number of warnings increases in PRs but doesn't care
about total number.

**Comment by mattip** [comment]
@pv it would only be an issue if this is merged with the failing tests. Subsequent pull requests based off master would fail lgtm tests and we wouldn't know the root cause

**Comment by eric-wieser** [comment]
Should I back out the LGTM CI?

**Comment by mattip** [comment]
It seems to be just this PR, perhaps we should give LGTM.com a chance to clarify what is going on

**Comment by pv** [comment]
I'd simply ignore CI failures that are the fault of the CI, so that they don't hold things back.

If it seems the CI cannot be easily fixed in reasonable time, then removing it is one option.

**Comment by samlanning** [comment]
Hi Everyone,

Sorry for the problems that LGTM is having with analysing this PR, and thank you for your patience.

I've dived into it a little more, and it does seem that the problem is submodules related. More specifically, the step in our analysis that fetches all of the required git commits for all submodules is failing on the introduction of the new sub-submodule `doc/sphinxext/doc/scipy-sphinx-theme` (it's currently assuming that it already exists, and tripping up).

I've passed this on to our development team and asked them how long it would take to implement a fix for this, and will get back to you as soon as I have an estimate for that.

**Comment by samlanning** [comment]
> This seems like a LGTM failure, even reverting the numpydoc submodule to 0.6 did not solve it. I am trying to work it out with them

Interesting, are you talking about https://github.com/mattip/numpy/commit/1be708ef8afd3d3d16849bc1fafef492ab53566d ? That did seem to fix the analysis as far as i can tell: https://lgtm.com/projects/g/numpy/numpy/rev/pr-88bba6b7638f9dd67caa69e362a3b0a1d4e6bbf1

**Comment by tibbes** [comment]
Sorry again for the problem with LGTM. It was caused by a bug analysing pull-requests that introduce a new submodule. We are still working on a permanent fix, but we have worked around the problem for this particular submodule so the lgtm PR checks are now green for this PR.

Also, be assured that merging this PR will _not_ cause future lgtm PR analyses to fail.

**Comment by mattip** [comment]
Thanks @tibbes

**Comment by mattip** [comment]
@rgommers the pr still has your "changes requested" review flag

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
LGTM now, merged. Thanks @mattip, all

**Pull request merged**